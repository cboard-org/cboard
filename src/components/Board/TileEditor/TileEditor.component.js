import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl, intlShape } from 'react-intl';
import shortid from 'shortid';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormLabel from '@material-ui/core/FormLabel';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import Paper from '@material-ui/core/Paper';
import TextField from '@material-ui/core/TextField';
import MobileStepper from '@material-ui/core/MobileStepper';
import Button from '@material-ui/core/Button';
import SearchIcon from '@material-ui/icons/Search';
import KeyboardArrowRightIcon from '@material-ui/icons/KeyboardArrowRight';
import KeyboardArrowLeftIcon from '@material-ui/icons/KeyboardArrowLeft';

import messages from './TileEditor.messages';
import SymbolSearch from '../SymbolSearch';
import Symbol from '../Symbol';
import Tile from '../Tile';
import FullScreenDialog, {
  FullScreenDialogContent
} from '../../UI/FullScreenDialog';
import InputImage from '../../UI/InputImage';
import IconButton from '../../UI/IconButton';
import ColorSelect from '../../UI/ColorSelect';
import './TileEditor.css';

export class TileEditor extends Component {
  static propTypes = {
    /**
     * @ignore
     */
    intl: intlShape.isRequired,
    /**
     * If true, TileEditor will be visibile
     */
    open: PropTypes.bool,
    /**
     * Callback fired on TileEditor request to be hidden
     */
    onClose: PropTypes.func.isRequired,
    /**
     * Tiles array to work on
     */
    editingTiles: PropTypes.array,
    /**
     * Callback fired when submitting edited board tiles
     */
    onEditSubmit: PropTypes.func.isRequired,
    /**
     * Callback fired when submitting a new board tile
     */
    onAddSubmit: PropTypes.func.isRequired
  };

  static defaultProps = {
    editingTiles: []
  };

  constructor(props) {
    super(props);

    this.defaultTileColors = {
      folder: '#bbdefb',
      symbol: '#fff176'
    };

    this.defaultTile = {
      label: '',
      labelKey: '',
      vocalization: '',
      image: '',
      loadBoard: '',
      backgroundColor: this.defaultTileColors.symbol
    };

    this.state = {
      activeStep: 0,
      editingTiles: props.editingTiles,
      isSymbolSearchOpen: false,
      selectedBackgroundColor: '',
      tile: this.defaultTile
    };
  }

  componentWillReceiveProps(props) {
    this.updateTileProperty('id', shortid.generate()); // todo not here
    this.setState({ editingTiles: props.editingTiles });
  }

  editingTile() {
    return this.state.editingTiles[this.state.activeStep];
  }

  currentTileProp(prop) {
    const currentTile = this.editingTile();
    return currentTile ? currentTile[prop] : this.state.tile[prop];
  }

  updateEditingTile(id, property, value) {
    return state => {
      const editingTiles = state.editingTiles.map(
        b => (b.id === id ? { ...b, ...{ [property]: value } } : b)
      );
      return { ...state, editingTiles };
    };
  }

  updateNewTile(property, value) {
    return state => {
      const tile = { ...state.tile, [property]: value };
      return { ...state, tile };
    };
  }

  updateTileProperty(property, value) {
    if (this.editingTile()) {
      this.setState(
        this.updateEditingTile(this.editingTile().id, property, value)
      );
    } else {
      this.setState(this.updateNewTile(property, value));
    }
  }

  handleSubmit = () => {
    const { onEditSubmit, onAddSubmit } = this.props;

    this.setState({
      activeStep: 0,
      selectedBackgroundColor: '',
      tile: this.defaultTile
    });

    if (this.editingTile()) {
      onEditSubmit(this.state.editingTiles);
    } else {
      const tileToAdd = this.state.tile;
      const selectedBackgroundColor = this.state.selectedBackgroundColor;

      if (selectedBackgroundColor) {
        tileToAdd.backgroundColor = selectedBackgroundColor;
      }
      onAddSubmit(tileToAdd);
    }
  };

  handleCancel = () => {
    const { onClose } = this.props;

    this.setState({
      activeStep: 0,
      selectedBackgroundColor: '',
      tile: this.defaultTile
    });
    onClose();
  };

  handleInputImageChange = image => {
    this.updateTileProperty('image', image);
  };

  handleSymbolSearchChange = ({ image, labelKey, label }) => {
    this.updateTileProperty('labelKey', labelKey);
    this.updateTileProperty('label', label);
    this.updateTileProperty('image', image);
  };

  handleSymbolSearchClose = event => {
    this.setState({ isSymbolSearchOpen: false });
  };

  handleLabelChange = event => {
    this.updateTileProperty('label', event.target.value);
    this.updateTileProperty('labelKey', '');
  };

  handleVocalizationChange = event => {
    this.updateTileProperty('vocalization', event.target.value);
  };

  handleTypeChange = (event, type) => {
    const typeFolder = type === 'folder';
    const loadBoard = typeFolder ? shortid.generate() : '';

    const backgroundColor = typeFolder
      ? this.defaultTileColors.folder
      : this.defaultTileColors.symbol;
    const tile = { ...this.state.tile, backgroundColor, loadBoard };
    this.setState({ tile });
  };

  handleBack = event => {
    this.setState({ activeStep: this.state.activeStep - 1 });
  };

  handleNext = event => {
    this.setState({ activeStep: this.state.activeStep + 1 });
  };

  handleSearchClick = event => {
    this.setState({ isSymbolSearchOpen: true });
  };

  handleColorChange = event => {
    const color = event ? event.target.value : '';
    this.setState({ selectedBackgroundColor: color });
  };

  getDefaultColor = () => {
    if (this.currentTileProp('loadBoard')) {
      return this.defaultTileColors.folder;
    }
    return this.defaultTileColors.symbol;
  };

  render() {
    const { open, intl } = this.props;

    const currentLabel = this.currentTileProp('labelKey')
      ? intl.formatMessage({ id: this.currentTileProp('labelKey') })
      : this.currentTileProp('label');

    const buttons = (
      <IconButton
        label={intl.formatMessage(messages.symbolSearch)}
        onClick={this.handleSearchClick}
      >
        <SearchIcon />
      </IconButton>
    );

    const tileInView = this.editingTile()
      ? this.editingTile()
      : this.state.tile;

    return (
      <div className="TileEditor">
        <FullScreenDialog
          disableSubmit={!currentLabel}
          buttons={buttons}
          open={open}
          title={
            <FormattedMessage
              {...(this.editingTile()
                ? messages.editTile
                : messages.createTile)}
            />
          }
          onClose={this.handleCancel}
          onSubmit={this.handleSubmit}
        >
          <Paper>
            <FullScreenDialogContent className="TileEditor__container">
              <div className="TileEditor__column">
                <div className="TileEditor__preview">
                  <Tile
                    backgroundColor={
                      this.state.selectedBackgroundColor ||
                      tileInView.backgroundColor
                    }
                    variant={
                      Boolean(tileInView.loadBoard) ? 'folder' : 'button'
                    }
                  >
                    <Symbol image={tileInView.image} label={currentLabel} />
                  </Tile>
                </div>
                <div className="TileEditor__input-image">
                  <InputImage onChange={this.handleInputImageChange} />
                </div>
              </div>
              <div className="TileEditor__fields">
                <TextField
                  id="label"
                  label={intl.formatMessage(messages.label)}
                  value={currentLabel}
                  onChange={this.handleLabelChange}
                  fullWidth
                  required
                />

                <TextField
                  id="vocalization"
                  label={intl.formatMessage(messages.vocalization)}
                  value={this.currentTileProp('vocalization') || ''}
                  onChange={this.handleVocalizationChange}
                  fullWidth
                />
                {!this.editingTile() && (
                  <div className="TileEditor__radiogroup">
                    <FormControl fullWidth>
                      <FormLabel>{intl.formatMessage(messages.type)}</FormLabel>
                      <RadioGroup
                        aria-label={intl.formatMessage(messages.type)}
                        name="type"
                        value={
                          this.currentTileProp('loadBoard')
                            ? 'folder'
                            : 'button'
                        }
                        onChange={this.handleTypeChange}
                      >
                        <FormControlLabel
                          value="button"
                          control={<Radio />}
                          label={intl.formatMessage(messages.button)}
                        />
                        <FormControlLabel
                          value="folder"
                          control={<Radio />}
                          label={intl.formatMessage(messages.folder)}
                        />
                      </RadioGroup>
                    </FormControl>
                    <ColorSelect
                      selectedColor={this.state.selectedBackgroundColor}
                      onChange={this.handleColorChange}
                    />
                  </div>
                )}
              </div>
            </FullScreenDialogContent>

            {this.state.editingTiles.length > 1 && (
              <MobileStepper
                variant="progress"
                steps={this.state.editingTiles.length}
                position="static"
                activeStep={this.state.activeStep}
                nextButton={
                  <Button
                    onClick={this.handleNext}
                    disabled={
                      this.state.activeStep ===
                      this.state.editingTiles.length - 1
                    }
                  >
                    {intl.formatMessage(messages.next)}{' '}
                    <KeyboardArrowRightIcon />
                  </Button>
                }
                backButton={
                  <Button
                    onClick={this.handleBack}
                    disabled={this.state.activeStep === 0}
                  >
                    <KeyboardArrowLeftIcon />
                    {intl.formatMessage(messages.back)}
                  </Button>
                }
              />
            )}
          </Paper>

          <SymbolSearch
            open={this.state.isSymbolSearchOpen}
            onChange={this.handleSymbolSearchChange}
            onClose={this.handleSymbolSearchClose}
          />
        </FullScreenDialog>
      </div>
    );
  }
}

export default injectIntl(TileEditor);
