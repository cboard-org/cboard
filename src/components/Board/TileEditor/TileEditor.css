.TileEditor__container {
  display: flex;
}

.TileEditor__column {
  width: 96px;
  flex-shrink: 0;
  margin-right: 16px;
}

@media (min-width: 600px) {
  .TileEditor__column {
    width: 144px;
  }
}

.TileEditor__fields {
  min-width: 0;
}

.TileEditor__preview {
  position: relative;
  margin-bottom: 10px;
}

.TileEditor__preview::before {
  content: '';
  display: block;
  padding-top: 100%;
}

.TileEditor__preview .Tile {
  position: absolute;
  top: 0;
  left: 0;
}

.TileEditor__radiogroup {
  margin-top: 15px;
}
