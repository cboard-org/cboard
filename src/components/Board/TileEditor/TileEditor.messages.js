import { defineMessages } from 'react-intl';

export default defineMessages({
  createTile: {
    id: 'cboard.components.Board.TileEditor.createTile',
    defaultMessage: 'Create tile'
  },
  editTile: {
    id: 'cboard.components.Board.TileEditor.editTile',
    defaultMessage: 'Edit Tile'
  },
  label: {
    id: 'cboard.components.Board.TileEditor.label',
    defaultMessage: 'Label'
  },
  vocalization: {
    id: 'cboard.components.Board.TileEditor.vocalization',
    defaultMessage: 'Vocalization'
  },
  button: {
    id: 'cboard.components.Board.TileEditor.button',
    defaultMessage: 'Button'
  },
  folder: {
    id: 'cboard.components.Board.TileEditor.folder',
    defaultMessage: 'Folder'
  },
  type: {
    id: 'cboard.components.Board.TileEditor.type',
    defaultMessage: 'Type'
  },
  back: {
    id: 'cboard.components.Board.TileEditor.back',
    defaultMessage: 'Back'
  },
  next: {
    id: 'cboard.components.Board.TileEditor.next',
    defaultMessage: 'Next'
  },
  symbolSearch: {
    id: 'cboard.components.Board.TileEditor.symbolSearch',
    defaultMessage: 'Symbol search'
  }
});
