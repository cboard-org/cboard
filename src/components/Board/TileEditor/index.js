export { default } from './TileEditor.component';
