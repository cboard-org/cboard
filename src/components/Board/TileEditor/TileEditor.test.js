import React from 'react';
import { mount, shallow } from 'enzyme';
import { matchSnapshotWithIntlProvider } from '../../../common/test_utils';

import TileEditor from './TileEditor.component';

jest.mock('./TileEditor.messages', () => {
  return {
    createTile: {
      id: 'cboard.components.Board.TileEditor.createTile',
      defaultMessage: 'Create tile'
    },
    editTile: {
      id: 'cboard.components.Board.TileEditor.editTile',
      defaultMessage: 'Edit Tile'
    },
    label: {
      id: 'cboard.components.Board.TileEditor.label',
      defaultMessage: 'Label'
    },
    vocalization: {
      id: 'cboard.components.Board.TileEditor.vocalization',
      defaultMessage: 'Vocalization'
    },
    voiceRecorder: {
      id: 'cboard.components.Board.TileEditor.voiceRecorder',
      defaultMessage: 'Voice Recorder'
    },
    button: {
      id: 'cboard.components.Board.TileEditor.button',
      defaultMessage: 'Button'
    },
    folder: {
      id: 'cboard.components.Board.TileEditor.folder',
      defaultMessage: 'Folder'
    },
    type: {
      id: 'cboard.components.Board.TileEditor.type',
      defaultMessage: 'Type'
    },
    back: {
      id: 'cboard.components.Board.TileEditor.back',
      defaultMessage: 'Back'
    },
    next: {
      id: 'cboard.components.Board.TileEditor.next',
      defaultMessage: 'Next'
    },
    symbolSearch: {
      id: 'cboard.components.Board.TileEditor.symbolSearch',
      defaultMessage: 'Symbol search'
    }
  };
});
describe('TileEditor tests', () => {
  const props = {
    intl: {
      formatMessage: jest.fn(),
      locale: 'en-US'
    },
    open: true,
    onClose: jest.fn(),
    editingTiles: [],
    onEditSubmit: jest.fn(),
    onAddSubmit: jest.fn()
  };

  test('default renderer', () => {
    matchSnapshotWithIntlProvider(<TileEditor {...props} />);
  });
  test('mount renderer', () => {
    const wrapper = shallow(<TileEditor {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
