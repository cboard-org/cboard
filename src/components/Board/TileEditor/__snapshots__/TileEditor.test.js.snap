// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileEditor tests default renderer 1`] = `null`;

exports[`TileEditor tests mount renderer 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TileEditor
    editingTiles={Array []}
    intl={
      Object {
        "formatMessage": [MockFunction],
        "locale": "en-US",
      }
    }
    onAddSubmit={[MockFunction]}
    onClose={[MockFunction]}
    onEditSubmit={[MockFunction]}
    open={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "editingTiles": Array [],
      "intl": Object {
        "formatMessage": [Function],
      },
      "onAddSubmit": [MockFunction],
      "onClose": [MockFunction],
      "onEditSubmit": [MockFunction],
      "open": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "editingTiles": Array [],
        "intl": Object {
          "formatMessage": [Function],
        },
        "onAddSubmit": [MockFunction],
        "onClose": [MockFunction],
        "onEditSubmit": [MockFunction],
        "open": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
